luigi/contrib/htcondor.py

```diff
@@ -0,0 +1,370 @@
+# limitations under the License.
+#
+from __future__ import absolute_import
+import shutil
+import stat
+import glob
+import time
+
+"""HTCondor batch system Tasks.
+
+Adapted by Luke Kreczko (@kreczko) from
+`SGE extension
+<https://github.com/spotify/luigi/blob/master/luigi/contrib/sge.py>`_
+by Jake Feala (@jfeala)
+
+HTCondor (http://research.cs.wisc.edu/htcondor/) is a job scheduler used to
+allocate compute resources on a shared cluster.
+Jobs are submitted and monitored using the HTCondor python bindings.
+To get started, install luigi on all nodes.
+
+To run luigi workflows on an HTCondor cluster, subclass
+:class:`luigi.contrib.htcondor.HTCondorJobTask` as you would any
+:class:`luigi.Task`, but override the ``work()`` method, instead of ``run()``,
+to define the job code. Then, run your Luigi workflow from the master node,
+assigning > 1 ``workers`` in order to distribute the tasks in parallel across
+the cluster.
+
+The following is an example usage (and can also be found in `htcondor_test.py`)
+
+.. code-block:: python
+
+    import logging
+    import luigi
+    import os
+    from luigi.contrib.sge import HTCondorJobTask
+
+    logger = logging.getLogger('luigi-interface')
+
+
+    class TestJobTask(HTCondorJobTask):
+
+        i = luigi.Parameter()
+
+        def work(self):
+            logger.info('Running test job...')
+            with open(self.output().path, 'w') as f:
+                f.write('this is a test')
+
+        def output(self):
+            return luigi.LocalTarget(
+                os.path.join('/tmp', 'testfile_' + str(self.i)))
+
+
+    if __name__ == '__main__':
+        tasks = [TestJobTask(i=str(i), n_cpu=i+1) for i in range(3)]
+        luigi.build(tasks, local_scheduler=True, workers=3)
+
+"""
+
+
+# This extension is modelled after the hadoop.py approach.
+#
+# Implementation notes
+# The procedure:
+# - Pickle the class
+# - Construct a htcondor job classad that runs a generic runner function with
+#   the path to the pickled class
+# - Runner function loads the class from pickle
+# - Runner function hits the work button on it
+
+import luigi
+import subprocess
+import os
+import logging
+import random
+import pwd
+import pickle
+import sys
+
+from luigi.contrib import sge_runner, hadoop
+
+logger = logging.getLogger('luigi-interface')
+logger.propagate = 0
+
+DEFAULT_JOB_PARAMETERS = {
+    'Universe': 'vanilla',
+    'request_cpus': 1,
+    'request_memory': 20,  # MB
+    'when_to_transfer_output': 'ON_EXIT_OR_EVICT',
+}
+
+RUN_SCRIPT = """#!/bin/bash
+BASE=${{_CONDOR_SCRATCH_DIR}}
+
+if [ -f $BASE/pre.sh ]; then
+  . $BASE/pre.sh
+fi
+
+python runner.py $BASE $BASE {tarball_param}
+
+if [ -f $BASE/post.sh ]; then
+  . $BASE/post.sh
+fi
+
+"""
+
+
+def _build_job_description(job_params):
+    from classad import ClassAd
+    submit_params = DEFAULT_JOB_PARAMETERS.copy()
+    submit_params.update(job_params)
+    return ClassAd(submit_params)
```

Comment:
I don't think you need to explicitly cast to `ClassAd` here; a dictionary should be fine on recent HTCondor versions.